### PR TITLE
Feature/mocks 181

### PR DESCRIPTION
**Pull request by AnnRiera**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/blockcoders/harmony-marketplace-sdk/blob/main/CONTRIBUTING.md
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## What is the new behavior?
Unit test for `HRC721`, `HRC1155` and `BaseToken`.

## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

File: src/hrc1155.spec.ts [hunk 1]
```diff
@@ -35,12 +36,14 @@ describe('HRC1155 Contract Interface', () => {
 
   describe.skip('balanceOf', () => {
     it('should get the number of tokens in the specified account with id as a number', async () => {
+      const mockedBalance = await FAKE_BALANCE_HRC1155
+      const stub = sinon.stub(contract, 'balanceOf').withArgs(TEST_ADDRESS_1, TOKEN_GOLD)
+      stub.resolves().returns(FAKE_BALANCE_HRC1155)
```
Comment (bzuker) [inline src/hrc1155.spec.ts:41]: here we are mocking the actual method being called, so this test won't test our functionality, but rather the sinon mock 😅. So, if we ever were to introduce a bug in our `balanceOf` method, we wouldn't catch it with the test.

What should be mocked is either the `call` method or the contract that lives inside the `BaseToken` instance.

File: src/hrc721.spec.ts [hunk 2]
```diff
@@ -35,158 +38,314 @@ describe('HRC721 Contract Interface', () => {
...
-      const balance = await contract.balanceOf(TEST_ADDRESS_1, TX_OPTIONS)
+      const mockedBalance = await FAKE_BALANCE_HRC721
+      const stub = sinon.stub(contract, 'call').withArgs('balanceOf', [TEST_ADDRESS_1], TX_OPTIONS)
+      stub.resolves().returns(FAKE_BALANCE_HRC721)
 
-      expect(balance).to.not.be.null
-      expect(balance).to.not.be.undefined
-      expect(balance).to.be.an.instanceof(BN)
-      expect(balance.gt(new BN(0))).to.be.true
+      const call = await contract.call<BN>('balanceOf', [TEST_ADDRESS_1], TX_OPTIONS)
```
Comment (bzuker) [inline src/hrc721.spec.ts:47]: we should use `balanceOf` directly, since that's the method we are testing..

File: src/hrc721.spec.ts [hunk 3]
```diff
@@ -35,158 +38,314 @@ describe('HRC721 Contract Interface', () => {
...
+      expect(stub.callCount).to.be.equals(1)
+      expect(call).to.be.equals(mockedBalance)
     })
 
     it('should throw an error if address is not provided', async () => {
-      expect(contract.balanceOf('')).to.be.rejectedWith(Error)
+      const stub = sinon.stub(contract, 'call')
+      stub.withArgs('balanceOf', [], TX_OPTIONS).onFirstCall().rejects()
+
+      expect(contract.call<BN>('balanceOf', [], TX_OPTIONS)).to.be.rejectedWith(Error)
```
Comment (bzuker) [inline src/hrc721.spec.ts:58]: same as above, here you are using `call` directly, but we should be testing the balanceOf method

File: src/hrc721.spec.ts [hunk 4]
```diff
@@ -35,158 +38,314 @@ describe('HRC721 Contract Interface', () => {
...
+    })
+
+    it('should throw an error if txOptions is not provided', async () => {
+      const stub = sinon.stub(contract, 'call')
+      stub.withArgs('balanceOf', [TEST_ADDRESS_1]).onFirstCall().rejects()
+
+      expect(contract.call<BN>('balanceOf', [TEST_ADDRESS_1])).to.be.rejectedWith(Error)
```
Comment (bzuker) [inline src/hrc721.spec.ts:65]: same as above

File: src/hrc721.spec.ts [hunk 5]
```diff
@@ -35,158 +38,314 @@ describe('HRC721 Contract Interface', () => {
...
     })
   })
 
-  describe.skip('ownerOf', () => {
+  describe('ownerOf', () => {
     it('should return the owner of the tokenId token', async () => {
-      const owner = await contract.ownerOf(TOKEN_GOLD, TX_OPTIONS)
+      const stub = sinon.stub(contract, 'call').withArgs('ownerOf', [TOKEN_GOLD], TX_OPTIONS)
+      stub.resolves().returns(FAKE_OWNER_HRC721)
+      const owner = await contract.call<string>('ownerOf', [TOKEN_GOLD], TX_OPTIONS)
```
Comment (bzuker) [inline src/hrc721.spec.ts:73]: here too

**Review by bzuker — CHANGES_REQUESTED** [review]
Left a few comments. The tests should test the actual method being called, instead of going around it.

I saw that in most cases you were using `call` or `send` in the tests, which would fail to actually test the method at hand, eg: 

when testing `transferFrom`, instead of doing `contract.transferFrom` you were using `contract.send("transferFrom")`, which won't test the method, but rather the `send` method. So any functionality that the `transferFrom` method has, won't be executed.

Let me know if you want to talk about this a bit more

File: src/hrc1155.spec.ts (hunk 1 above)
Comment (AnnRiera) [reply src/hrc1155.spec.ts:41]: You're right!

File: src/hrc721.spec.ts (hunk 2 above)
Comment (AnnRiera) [reply src/hrc721.spec.ts:47]: Got it!

File: src/hrc721.spec.ts (hunk 3 above)
Comment (AnnRiera) [reply src/hrc721.spec.ts:58]: Got it!

File: src/hrc721.spec.ts (hunk 4 above)
Comment (AnnRiera) [reply src/hrc721.spec.ts:65]: Got it!

File: src/hrc721.spec.ts (hunk 5 above)
Comment (AnnRiera) [reply src/hrc721.spec.ts:73]: Copy that!

File: src/tests/constants.ts [hunk 6]
```diff
@@ -57,3 +56,38 @@ export const TX_OPTIONS: ITransactionOptions = {
...
+      toShardID: 0,
+      nonce: 165,
+      chainId: 2,
+    }),
+  )
+})
+
+export const FAKE_BALANCE_HRC1155 = new Promise<BN>((resolve) => {
+  resolve(new BN('999999999999999999999889'))
+})
```
Comment (bzuker) [inline src/tests/constants.ts:93]: why are these promises? I think they could just be the result to what it resolves

File: src/hrc1155.spec.ts [hunk 7]
```diff
@@ -33,178 +33,313 @@ describe('HRC1155 Contract Interface', () => {
...
-      const balance = await contract.balanceOf(TEST_ADDRESS_1, TOKEN_GOLD)
+      const mockedBalance = await FAKE_BALANCE_HRC1155
+      const stub = sinon.stub(contract, 'call').withArgs('balanceOf', [TEST_ADDRESS_1, TOKEN_GOLD], TX_OPTIONS)
+      stub.resolves().returns(FAKE_BALANCE_HRC1155)
 
-      expect(balance).to.not.be.null
-      expect(balance).to.not.be.undefined
-      expect(balance).to.be.an.instanceof(BN)
-      expect(balance.gt(new BN(0))).to.be.true
+      const call = await contract.balanceOf(TEST_ADDRESS_1, TOKEN_GOLD, TX_OPTIONS)
```
Comment (bzuker) [inline src/hrc1155.spec.ts:42]: nit: I'd call this `result` or `balance` so that it's clearer. Similar in other tests

File: src/hrc1155.spec.ts [hunk 8]
```diff
@@ -33,178 +33,313 @@ describe('HRC1155 Contract Interface', () => {
...
-      expect(balance.gt(new BN(0))).to.be.true
+      expect(stub.calledOnce).to.be.true
+      expect(stub.callCount).to.be.equals(1)
+      expect(call).to.be.equals(mockedBalance)
     })
 
     it('should throw an error if address is not provided', async () => {
-      expect(contract.balanceOf('', 1)).to.be.rejectedWith(Error)
+      const stub = sinon.stub(contract, 'call')
+      stub.withArgs('balanceOf', ['', TOKEN_GOLD], TX_OPTIONS).onFirstCall().rejects()
```
Comment (bzuker) [inline src/hrc1155.spec.ts:65]: probably not necessary to mock here

File: src/hrc1155.spec.ts [hunk 9]
```diff
@@ -33,178 +33,313 @@ describe('HRC1155 Contract Interface', () => {
...
+
+      expect(contract.balanceOf('', TOKEN_GOLD, TX_OPTIONS)).to.be.rejectedWith(Error)
     })
 
     it('should throw an error if id is not provided', async () => {
-      expect(contract.balanceOf(TEST_ADDRESS_1, 0)).to.be.rejectedWith(Error)
+      const stub = sinon.stub(contract, 'call')
+      stub.withArgs('balanceOf', [TEST_ADDRESS_1, ''], TX_OPTIONS).onFirstCall().rejects()
```
Comment (bzuker) [inline src/hrc1155.spec.ts:72]: same as above

File: src/hrc1155.spec.ts [hunk 10]
```diff
@@ -33,178 +33,313 @@ describe('HRC1155 Contract Interface', () => {
...
+
+      expect(contract.balanceOf(TEST_ADDRESS_1, '', TX_OPTIONS)).to.be.rejectedWith(Error)
     })
 
     it('should throw an error if params are not provided', async () => {
-      expect(contract.balanceOf('', 0)).to.be.rejectedWith(Error)
+      const stub = sinon.stub(contract, 'call')
+      stub.withArgs('balanceOf', ['', '']).onFirstCall().rejects()
```
Comment (bzuker) [inline src/hrc1155.spec.ts:79]: same as above

File: src/hrc1155.spec.ts [hunk 11]
```diff
@@ -33,178 +33,313 @@ describe('HRC1155 Contract Interface', () => {
...
+      expect(stub.callCount).to.be.equals(1)
+      expect(call).to.be.equals(mockedBalance)
     })
 
     it('should throw an error if ids is not provided', async () => {
+      const stub = sinon.stub(contract, 'call')
+      stub
+        .withArgs('balanceOfBatch', [[], [1, 2]])
+        .onFirstCall()
+        .rejects()
```
Comment (bzuker) [inline src/hrc1155.spec.ts:160]: probably not necessary

File: src/hrc1155.spec.ts [hunk 12]
```diff
@@ -33,178 +33,313 @@ describe('HRC1155 Contract Interface', () => {
...
+
       expect(contract.balanceOfBatch([], [1, 2])).to.be.rejectedWith(Error)
     })
 
     it('should throw an error if accounts is not provided', async () => {
+      const stub = sinon.stub(contract, 'call')
+      stub
+        .withArgs('balanceOfBatch', [[TEST_ADDRESS_1, EMPTY_TEST_ADDRESS], []])
+        .onFirstCall()
+        .rejects()
```
Comment (bzuker) [inline src/hrc1155.spec.ts:170]: same

File: src/hrc1155.spec.ts [hunk 13]
```diff
@@ -33,178 +33,313 @@ describe('HRC1155 Contract Interface', () => {
...
+
       expect(contract.balanceOfBatch([TEST_ADDRESS_1, EMPTY_TEST_ADDRESS], [])).to.be.rejectedWith(Error)
     })
 
     it('should throw an error if params are not provided', async () => {
+      const stub = sinon.stub(contract, 'call')
+      stub.withArgs('balanceOfBatch', [[], []]).onFirstCall().rejects()
```
Comment (bzuker) [inline src/hrc1155.spec.ts:177]: same

**Review by bzuker — APPROVED** [review]
Some more comments, but looking good! 
Check that when testing for errors (most of the `should throw...`) the mocks are probably not necessary. Once those are removed, LGTM!

File: src/hrc1155.spec.ts (hunk 7 above)
Comment (0xslipk) [reply src/hrc1155.spec.ts:42]: agree!

File: src/tests/constants.ts (hunk 6 above)
Comment (0xslipk) [reply src/tests/constants.ts:93]: yeah I think so

**Review by 0xslipk — COMMENTED** [review]
@AnnRiera please address Brian's comments

File: src/hrc1155.spec.ts (hunk 7 above)
Comment (AnnRiera) [reply src/hrc1155.spec.ts:42]: Okay!

File: src/hrc1155.spec.ts (hunk 8 above)
Comment (AnnRiera) [reply src/hrc1155.spec.ts:65]: Copy

File: src/hrc1155.spec.ts (hunk 9 above)
Comment (AnnRiera) [reply src/hrc1155.spec.ts:72]: Copy

File: src/hrc1155.spec.ts (hunk 10 above)
Comment (AnnRiera) [reply src/hrc1155.spec.ts:79]: Copy

File: src/hrc1155.spec.ts (hunk 11 above)
Comment (AnnRiera) [reply src/hrc1155.spec.ts:160]: Copy

File: src/hrc1155.spec.ts (hunk 12 above)
Comment (AnnRiera) [reply src/hrc1155.spec.ts:170]: Copy

File: src/hrc1155.spec.ts (hunk 13 above)
Comment (AnnRiera) [reply src/hrc1155.spec.ts:177]: Copy

File: src/tests/constants.ts (hunk 6 above)
Comment (AnnRiera) [reply src/tests/constants.ts]: Got it!

**Review by bzuker — APPROVED** [review]
LGTM!

**Review by 0xslipk — APPROVED** [review]

**Pull request merged**